File: proj.android_studio/ee-x-app-lovin/src/main/java/com/ee/AppLovinBridge.kt
package com.ee

import android.app.Activity
import android.content.Context
import androidx.annotation.AnyThread
import com.applovin.adview.AppLovinIncentivizedInterstitial
import com.applovin.adview.AppLovinInterstitialAd
import com.applovin.adview.AppLovinInterstitialAdDialog
import com.applovin.sdk.AppLovinAdSize
import com.applovin.sdk.AppLovinSdk
import com.applovin.sdk.AppLovinSdkSettings
import com.ee.internal.AppLovinInterstitialAdListener
import com.ee.internal.AppLovinRewardedAdListener

class AppLovinBridge(
    private val _bridge: IMessageBridge,
    private val _context: Context,
    private var _activity: Activity?) : IPlugin {
    companion object {
        private val _logger = Logger(AppLovinBridge::class.java.name)

        private const val kPrefix = "AppLovinBridge"
        private const val kInitialize = "${kPrefix}Initialize"
        private const val kSetVerboseLogging = "${kPrefix}SetVerboseLogging"
        private const val kSetMuted = "${kPrefix}SetMuted"
        private const val kHasInterstitialAd = "${kPrefix}HasInterstitialAd"
        private const val kLoadInterstitialAd = "${kPrefix}LoadInterstitialAd"
        private const val kShowInterstitialAd = "${kPrefix}ShowInterstitialAd"
        private const val kHasRewardedAd = "${kPrefix}HasRewardedAd"
        private const val kLoadRewardedAd = "${kPrefix}LoadRewardedAd"
        private const val kShowRewardedAd = "${kPrefix}ShowRewardedAd"
    }

    private var _initializing = false
    private var _initialized = false
    private var _sdk: AppLovinSdk? = null
    private var _interstitialAd: AppLovinInterstitialAdDialog? = null
    private var _interstitialAdListener: AppLovinInterstitialAdListener? = null
    private var _rewardedAd: AppLovinIncentivizedInterstitial? = null
    private var _rewardedAdListener: AppLovinRewardedAdListener? = null

    init {
        _logger.debug("constructor begin: context = $_context")
        registerHandlers()
        _logger.debug("constructor end.")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
    }

    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {}
    override fun onPause() {}
    override fun onDestroy() {
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
        Thread.runOnMainThread {
            if (!_initialized) {
                return@runOnMainThread
            }
            _sdk = null
            _interstitialAd?.let { ad ->
                ad.setAdLoadListener(null)
                ad.setAdDisplayListener(null)
                ad.setAdClickListener(null)
            }
            _interstitialAd = null
            _interstitialAdListener = null
            _rewardedAd = null
            _rewardedAdListener = null
        }
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(kInitialize) { message ->
            initialize(message)
            ""
        }
        _bridge.registerHandler(kSetVerboseLogging) { message ->
            setVerboseLogging(Utils.toBoolean(message))
            ""
        }
        _bridge.registerHandler(kSetMuted) { message ->
            setMuted(Utils.toBoolean(message))
            ""
        }
        _bridge.registerHandler(kHasInterstitialAd) {
            Utils.toString(hasInterstitialAd)
        }
        _bridge.registerHandler(kLoadInterstitialAd) {
            loadInterstitialAd()
            ""
        }
        _bridge.registerHandler(kShowInterstitialAd) {
            showInterstitialAd()
            ""
        }
        _bridge.registerHandler(kHasRewardedAd) {
            Utils.toString(hasRewardedAd)
        }
        _bridge.registerHandler(kLoadRewardedAd) {
            loadRewardedAd()
            ""
        }
        _bridge.registerHandler(kShowRewardedAd) {
            showRewardedAd()
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kSetVerboseLogging)
        _bridge.deregisterHandler(kSetMuted)
        _bridge.deregisterHandler(kHasInterstitialAd)
        _bridge.deregisterHandler(kLoadInterstitialAd)
        _bridge.deregisterHandler(kShowInterstitialAd)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kLoadRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
    }

    @AnyThread
    fun initialize(key: String) {
        Thread.runOnMainThread {
            if (_initializing) {
                return@runOnMainThread
            }
            if (_initialized) {
                return@runOnMainThread
            }
            _initializing = true
            val settings = AppLovinSdkSettings()
            val sdk = AppLovinSdk.getInstance(key, settings, _context)
            sdk.initializeSdk {
                _initializing = false
                _initialized = true
            }

            val interstitialAdListener = AppLovinInterstitialAdListener(_bridge)
            val interstitialAd = AppLovinInterstitialAd.create(sdk, _activity)
            interstitialAd.setAdLoadListener(interstitialAdListener)
            interstitialAd.setAdDisplayListener(interstitialAdListener)
            interstitialAd.setAdClickListener(interstitialAdListener)

            val rewardedAd = AppLovinIncentivizedInterstitial.create(sdk)
            val rewardedAdListener = AppLovinRewardedAdListener(_bridge)

            _sdk = sdk
            _interstitialAd = interstitialAd
            _interstitialAdListener = interstitialAdListener
            _rewardedAd = rewardedAd
            _rewardedAdListener = rewardedAdListener
        }
    }

    @AnyThread
    fun setVerboseLogging(enabled: Boolean) {
        Thread.runOnMainThread {
            val sdk = _sdk ?: throw IllegalStateException("Please call initialize() first")
            sdk.settings.setVerboseLogging(enabled)
        }
    }

    @AnyThread
    fun setMuted(enabled: Boolean) {
        Thread.runOnMainThread {
            val sdk = _sdk ?: throw IllegalStateException("Please call initialize() first")
            sdk.settings.isMuted = enabled
        }
    }

    private val hasInterstitialAd: Boolean
        @AnyThread get() = _interstitialAdListener?.isLoaded ?: false

    @AnyThread
    fun loadInterstitialAd() {
        Thread.runOnMainThread {
            val sdk = _sdk ?: throw IllegalStateException("Please call initialize() first")
            sdk.adService.loadNextAd(AppLovinAdSize.INTERSTITIAL, _interstitialAdListener)
        }
    }

    @AnyThread
    fun showInterstitialAd() {
        Thread.runOnMainThread {
            val ad = _interstitialAd
                ?: throw IllegalStateException("Please call initialize() first")
            ad.show()
        }
    }

    val hasRewardedAd: Boolean
        @AnyThread get() = _rewardedAdListener?.isLoaded ?: false

    @AnyThread
    fun loadRewardedAd() {
        Thread.runOnMainThread {
            val ad = _rewardedAd ?: throw IllegalStateException("Please call initialize() first")
            ad.preload(_rewardedAdListener)
        }
    }

    @AnyThread
    fun showRewardedAd() {
        Thread.runOnMainThread {
            val ad = _rewardedAd ?: throw IllegalStateException("Please call initialize() first")
            ad.show(_context, _rewardedAdListener, _rewardedAdListener, _rewardedAdListener, _rewardedAdListener)
        }
    }
}

File: src/ios/ee/app_lovin/AppLovinBridge.swift
//
//  AppLovinBridge.swift
//  Pods
//
//  Created by eps on 6/26/20.
//

import AppLovinSDK
import RxSwift

private let kPrefix = "AppLovinBridge"
private let kInitialize = "\(kPrefix)Initialize"
private let kSetVerboseLogging = "\(kPrefix)SetVerboseLogging"
private let kSetMuted = "\(kPrefix)SetMuted"
private let kHasInterstitialAd = "\(kPrefix)HasInterstitialAd"
private let kLoadInterstitialAd = "\(kPrefix)LoadInterstitialAd"
private let kShowInterstitialAd = "\(kPrefix)ShowInterstitialAd"
private let kHasRewardedAd = "\(kPrefix)HasRewardedAd"
private let kLoadRewardedAd = "\(kPrefix)LoadRewardedAd"
private let kShowRewardedAd = "\(kPrefix)ShowRewardedAd"

@objc(EEAppLovinBridge)
class AppLovinBridge: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _initializing = false
    private var _initialized = false
    private var _sdk: ALSdk?
    private var _interstitialAd: ALInterstitialAd?
    private var _interstitialAdListener: AppLovinInterstitialAdListener?
    private var _rewardedAd: ALIncentivizedInterstitialAd?
    private var _rewardedAdListener: AppLovinRewardedAdListener?
    
    public required init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
        registerHandlers()
    }
    
    public func destroy() {
        deregisterHandlers()
        Thread.runOnMainThread {
            if !self._initialized {
                return
            }
            self._sdk = nil
            self._interstitialAd?.adLoadDelegate = nil
            self._interstitialAd?.adDisplayDelegate = nil
            self._interstitialAd = nil
            self._rewardedAd?.adDisplayDelegate = nil
            self._rewardedAd?.adVideoPlaybackDelegate = nil
            self._rewardedAd = nil
        }
    }
    
    func registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { message, resolver in
            self.initialize(message)
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
        _bridge.registerHandler(kSetVerboseLogging) { message in
            self.setVerboseLogging(Utils.toBool(message))
            return ""
        }
        _bridge.registerHandler(kSetMuted) { message in
            self.setMuted(Utils.toBool(message))
            return ""
        }
        _bridge.registerHandler(kHasInterstitialAd) { _ in
            Utils.toString(self.hasInterstitialAd)
        }
        _bridge.registerHandler(kLoadInterstitialAd) { _ in
            self.loadInterstitialAd()
            return ""
        }
        _bridge.registerHandler(kShowInterstitialAd) { _ in
            self.showInterstitialAd()
            return ""
        }
        _bridge.registerHandler(kHasRewardedAd) { _ in
            Utils.toString(self.hasRewardedAd)
        }
        _bridge.registerHandler(kLoadRewardedAd) { _ in
            self.loadRewardedAd()
            return ""
        }
        _bridge.registerHandler(kShowRewardedAd) { _ in
            self.showRewardedAd()
            return ""
        }
    }
    
    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kSetVerboseLogging)
        _bridge.deregisterHandler(kSetMuted)
        _bridge.deregisterHandler(kHasInterstitialAd)
        _bridge.deregisterHandler(kLoadInterstitialAd)
        _bridge.deregisterHandler(kShowInterstitialAd)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kLoadRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
    }
    
    func checkInitialized() {
        if !_initialized {
            assert(false, "Please call initialize() first")
        }
    }
    
    func initialize(_ key: String) -> Single<Bool> {
        return Single<Bool>.create { single in
            Thread.runOnMainThread {
                if self._initializing {
                    single(.success(false))
                    return
                }
                if self._initialized {
                    single(.success(true))
                    return
                }
                self._initializing = true
                guard let sdk = ALSdk.shared(withKey: key) else {
                    assert(false, "Invalid key")
                    single(.success(false))
                    return
                }
                sdk.initializeSdk(completionHandler: { _ in
                    Thread.runOnMainThread {
                        self._initializing = false
                        self._initialized = true
                        single(.success(true))
                    }
                })
                
                let interstitialAdListener = AppLovinInterstitialAdListener(self._bridge, self._logger)
                let interstitialAd = ALInterstitialAd(sdk: sdk)
                interstitialAd.adLoadDelegate = interstitialAdListener
                interstitialAd.adDisplayDelegate = interstitialAdListener
                
                let rewardedAdListener = AppLovinRewardedAdListener(self._bridge, self._logger)
                let rewardedAd = ALIncentivizedInterstitialAd(sdk: sdk)
                rewardedAd.adDisplayDelegate = rewardedAdListener
                rewardedAd.adVideoPlaybackDelegate = rewardedAdListener
                
                self._sdk = sdk
                self._interstitialAd = interstitialAd
                self._interstitialAdListener = interstitialAdListener
                self._rewardedAd = rewardedAd
                self._rewardedAdListener = rewardedAdListener
            }
            return Disposables.create()
        }
        .subscribeOn(MainScheduler())
    }
    
    func setVerboseLogging(_ enabled: Bool) {
        Thread.runOnMainThread {
            self.checkInitialized()
            guard let sdk = self._sdk else { return }
            sdk.settings.isVerboseLoggingEnabled = enabled
        }
    }
    
    func setMuted(_ enabled: Bool) {
        Thread.runOnMainThread {
            self.checkInitialized()
            guard let sdk = self._sdk else { return }
            sdk.settings.isMuted = enabled
        }
    }
    
    var hasInterstitialAd: Bool {
        return _interstitialAdListener?.isLoaded ?? false
    }
    
    func loadInterstitialAd() {
        Thread.runOnMainThread {
            self.checkInitialized()
            guard
                let sdk = self._sdk,
                let listener = self._interstitialAdListener
            else { return }
            sdk.adService.loadNextAd(ALAdSize.interstitial, andNotify: listener)
        }
    }
    
    func showInterstitialAd() {
        Thread.runOnMainThread {
            self.checkInitialized()
            guard let ad = self._interstitialAd else { return }
            ad.show()
        }
    }
    
    var hasRewardedAd: Bool {
        if !_initialized {
            return false
        }
        return _rewardedAdListener?.isLoaded ?? false
    }
    
    func loadRewardedAd() {
        Thread.runOnMainThread {
            self.checkInitialized()
            guard let ad = self._rewardedAd else { return }
            ad.preloadAndNotify(self._rewardedAdListener)
        }
    }
    
    func showRewardedAd() {
        Thread.runOnMainThread {
            self.checkInitialized()
            guard let ad = self._rewardedAd else { return }
            ad.showAndNotify(self._rewardedAdListener)
        }
    }
}
